# install Tensorflow

In [2]:
import tensorflow as tf

In [4]:
# check version
tf.__version__

'1.14.0'

 # Tensorflow Hello World!

In [5]:
# create a constant op
# This op is added as a node to the default graph
hello = tf.constant("Hello, Tensorflow") # 실제로는 그래프 속의 노드가 hello이다. 

# start a TF session
sess = tf.Session()

# run the op and get result
print(sess.run(hello))

# Here, b indicates Byte literals.
# node를 만들고 실행시켰다. 

b'Hello, Tensorflow'


# Computational Graph

Node가 2개 있고, 더하기가 있는 노드와 연결이 되는 간단한 그래프이다. 

In [6]:
node1 = tf.constant(3.0, tf.float32)
node2 = tf.constant(4.0)
node3 = tf.add(node1, node2) #혹은, node3 = node1 + node2

In [7]:
# 결과 값이 나오지 않는다. 
print("node1: ", node1, "node2: ",node2)
print("node3: ", node3)

node1:  Tensor("Const_1:0", shape=(), dtype=float32) node2:  Tensor("Const_2:0", shape=(), dtype=float32)
node3:  Tensor("Add:0", shape=(), dtype=float32)


In [9]:
# 실제 결과값을 보기 위해선, session을 만들고 run해주어야 한다. 
sess = tf.Session()
print("sess.run(node1, node2): ", sess.run([node1, node2]))
print("sess.run(node3): ", sess.run(node3))

sess.run(node1, node2):  [3.0, 4.0]
sess.run(node3):  7.0


1. 그래프를 build하고

2. session으로 run 한다. 

3. 그래프에서 값을 update한다. 

 # Placeholder
 
만약 그래프를 만들고, 실행시키는 단계에서 값을 주고 싶으면 어떻게 할것인가?
 
이럴 때도 node를 만드는데, placeholder라는 node를 만든다. 

In [10]:
a = tf.placeholder(tf.float32)
b = tf.placeholder(tf.float32)
adder_node = a + b 

In [11]:
# 실제 실행을 할 때, 값을 넘겨준다. 
print(sess.run(adder_node, feed_dict={a:3, b:4.5}))

# 넘겨주고 싶은 동시에 여러개 이면 아래와 같이 넘겨줄수도 있다.
print(sess.run(adder_node, feed_dict={a:[1,3], b:[2,4]}))

7.5
[3. 7.]


1. 그래프를 정의한다. ex) placeholder
2. session을 통해 실행시킬 때, feed_dict으로 값을 넘겨줘야 한다. 

# Everything is Tensor

Tensor라는 것은 기본적으로  array이다. 

### Ranks
rank가 0이면, Scalar, ex) s = 483
rank가 1이면, Vector, ex) v = [1.1, 2.2, 3.3]
rank가 2이면, Matrix, ex) m = [[1,2,3],[4,5,6],[7,8,9]]
rank가 3이면, 3-Tensor, ex) t = [[[2],[4],[6]],[[8],[10],[12]]]
rank가 n이면, n-Tensor, .....


### Shapes
차원에 따라 shape이 다르다. 
[0],[0,0], [0,0,0].....

### Types
주로 flaot32를 사용한다. 